### PR TITLE
Automated Content View test - test_cv_clone_within_same_env closes #2076

### LINKED ISSUE #2076: Automate `test_cv_clone_within_same_env`
This test is around cv clone functionality...

### PR DESCRIPTION
**Pull request by jyejare**
Automated test 'test_cv_clone_within_same_env'.

Passing logs:
```
[root@jyejare robottelo]# nosetests -m test_cv_clone_within_same_env ./tests/foreman/ui/test_contentviews.py
.
----------------------------------------------------------------------
Ran 1 test in 155.509s

OK
```

### CODE REVIEW

**Comment by sghai** [comment]
@jyejare : travis is throwing some error.. please fix them..

```
./robottelo/ui/contentviews.py:12:1: F401 'Element' imported but unused
./robottelo/ui/contentviews.py:13:1: F401 'NoSuchElementException' imported but unused
```

**Comment by jyejare** [comment]
Fixed in above push. Removed those unused import libraries as per flake8.

File: robottelo/ui/contentviews.py [hunk 1]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
             else:
                 raise UINoSuchElementError(
                     "Couldn't find puppet-modules tab")
+
+    def copy_view(self, name, new_name=None):
+        """
+        Copies an existing Content View
+        """
+        cv = self.search(name)
+
```
Comment (sghai) [inline robottelo/ui/contentviews.py]: Please remove this blank line..I think its not required here...

Comment (jyejare) [inline robottelo/ui/contentviews.py]: Cool ... I will remove.

**Comment by sghai** [comment]
ACK.. Nicely done..

**Comment by jyejare** [comment]
@sghai : Worked on all your comments and update the code.

**Comment by Ichimonji10** [comment]
Mr. Style says:

Please don't split dict lookups across two lines. You'll see that done here and there in the code base still, but I'd like us to move away from that.

```python
self.wait_until_element(
    locators['contentviews.yum_repositories']
).click()
```

Please use single qutoes instead of double quotes. It's a minor thing, but consistently using one kind of quotes allows the reader to skim through the code with one less distraction. A fundamental law of programming is that it's harder to read code than to write it, and I'd like to ease that burden as much as possible.

Use "sphinx-style" docstrings. This is preferred:

    """There was an old man of Tarentum,"""
    pass

This is not:

    """
    who gnashed his false teeth 'till he bent 'em.
    """
    pass

Similarly, this is preferred:

    """When asked the cost of what he had lost,

    he replied,

    "I can't say, for I rent 'em."

    """
    pass

Avoid hanging code, and place one argument on each line. Again, this is done in an attempt to make code as uniform - and therefore legible - as possible. This is discouraged:

```python
self.edit_entity(locators['contentviews.copy'],
                locators['contentviews.copy_name'],
                new_name, locators['ak.copy_create'])
```

This is preferred

```python
self.edit_entity(
    locators['contentviews.copy'],
    locators['contentviews.copy_name'],
    new_name,
    locators['ak.copy_create'],
)
```

File: robottelo/ui/contentviews.py [hunk 2]
```diff
@@ -722,3 +722,40 @@ def fetch_puppet_module(self, cv_name, module_name):
             else:
                 raise UINoSuchElementError(
                     "Couldn't find puppet-modules tab")
+
+    def copy_view(self, name, new_name=None):
+        """
+        Copies an existing Content View
+        """
+        cv = self.search(name)
+        if cv and new_name:
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: `search` returns `None` if it fails to find an element. Performing an explicit comparison against `None` is more robust and will prevent subtle bugs.

    if cv is not None and new_name is not None:

File: robottelo/ui/contentviews.py [hunk 3]
```diff
@@ -722,3 +722,40 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        else:
+            raise UIError('Could not copy the Content View %s .', name)
+
+    def fetch_yum_content_repo_name(self, cv_name):
+        """
+        Fetch associated content host from selected activation key
+        """
+        # find content_view
+        cv = self.search(cv_name)
+        if not cv:
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Performing an explicit comparison against `None` may prevent some subtle bugs.

    if cv is None

**Comment by Ichimonji10** [comment]
ACK pending comments.

File: tests/foreman/ui/test_contentviews.py [hunk 4]
```diff
@@ -977,6 +976,27 @@ def test_cv_clone_within_same_env(self):
...
+        cv_name = gen_string('alpha', 8)
+        copy_cv_name = gen_string('alpha', 8)
+        with Session(self.browser) as session:
+            self.setup_to_create_cv(repo_name=repo_name)
+            # Create content-view
+            make_contentview(session, org=self.org_name, name=cv_name)
+            self.assertIsNotNone(self.content_views.search(cv_name))
+            # Add repository to selected CV
+            self.content_views.add_remove_repos(cv_name, [repo_name])
+            self.assertIsNotNone(self.content_views.wait_until_element
```
Comment (elyezer) [inline tests/foreman/ui/test_contentviews.py]: This should look better

```python
self.assertIsNotNone(
    self.content_views.wait_until_element(
        common_locators["alert.success"])
)
```

Having the code as left aligned as possible increase legibility

File: tests/foreman/ui/test_contentviews.py [hunk 5]
```diff
@@ -977,6 +976,27 @@ def test_cv_clone_within_same_env(self):
...
+                                 (common_locators["alert.success"]))
+            # Publish the CV
+            self.content_views.publish(cv_name)
+            self.assertIsNotNone(self.content_views.wait_until_element
```
Comment (elyezer) [inline tests/foreman/ui/test_contentviews.py]: Ditto

File: tests/foreman/ui/test_contentviews.py [hunk 6]
```diff
@@ -977,6 +976,27 @@ def test_cv_clone_within_same_env(self):
...
+                                 (common_locators["alert.success"]))
+            # Copy the CV
+            self.content_views.copy_view(cv_name, copy_cv_name)
+            self.assertIsNotNone(self.content_views.search(copy_cv_name))
+            self.assertEqual(repo_name, self.content_views.
```
Comment (elyezer) [inline tests/foreman/ui/test_contentviews.py]: Same as https://github.com/SatelliteQE/robottelo/pull/2079/files#r28529553

**Comment by elyezer** [comment]
ACK, pending Mr. Style comments :smile:

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Updated code accordingly !!

File: robottelo/ui/contentviews.py (hunk 3 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Thanks ... updated.

File: tests/foreman/ui/test_contentviews.py (hunk 4 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Done.

File: tests/foreman/ui/test_contentviews.py (hunk 5 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Thanks for comments. Done.

File: tests/foreman/ui/test_contentviews.py (hunk 6 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Updated. Thanks.

**Comment by jyejare** [comment]
All the comments to improve the code has been taken care. Please review it again. Also thanks for all the nice comments @Ichimonji10 and @elyezer .

File: robottelo/ui/contentviews.py [hunk 7]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        """Fetch associated content host from selected activation key."""
+        # find content_view
+        cv = self.search(cv_name)
+        if cv is None:
+            raise UINoSuchElementError(
+                'Could not find CV %s', cv_name)
+        cv.click()
+        self.wait_for_ajax()
+        self.wait_until_element(
+            tab_locators["contentviews.tab_content"]
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Nitpick: `s/"/'/g`

File: robottelo/ui/contentviews.py [hunk 8]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        ).click()
+        self.wait_until_element(
+            locators["contentviews.yum_repositories"]
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Nitpick: `s/"/'/g`

File: robottelo/ui/contentviews.py [hunk 9]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        ).click()
+        self.wait_for_ajax()
+        if self.wait_until_element(locators["contentviews.repo_name"]):
+            result = self.find_element(locators["contentviews.repo_name"]).text
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Nitpick: `s/"/'/g`

File: robottelo/ui/contentviews.py [hunk 10]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        cv.click()
+        self.wait_for_ajax()
+        self.wait_until_element(
+            tab_locators["contentviews.tab_content"]
+        ).click()
+        self.wait_until_element(
+            locators["contentviews.yum_repositories"]
+        ).click()
+        self.wait_for_ajax()
+        if self.wait_until_element(locators["contentviews.repo_name"]):
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Nitpick: `s/"/'/g`

File: robottelo/ui/contentviews.py [hunk 11]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+            result = self.find_element(locators["contentviews.repo_name"]).text
+            return result
+        else:
+            raise UINoSuchElementError(
+                "Couldn't get text attribute of repository locator")
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Nitpick: `s/"/'/g`

File: tests/foreman/ui/test_contentviews.py [hunk 12]
```diff
@@ -977,6 +976,37 @@ def test_cv_clone_within_same_env(self):
...
+            self.content_views.add_remove_repos(cv_name, [repo_name])
+            self.assertIsNotNone(
+                self.content_views.wait_until_element(
+                    common_locators["alert.success"])
+            )
+            # Publish the CV
+            self.content_views.publish(cv_name)
+            self.assertIsNotNone(
+                self.content_views.wait_until_element(
+                    common_locators["alert.success"])
```
Comment (Ichimonji10) [inline tests/foreman/ui/test_contentviews.py]: Nitpick: `s/"/'/g`

File: tests/foreman/ui/test_contentviews.py [hunk 13]
```diff
@@ -977,6 +976,37 @@ def test_cv_clone_within_same_env(self):
...
+            # Create content-view
+            make_contentview(session, org=self.org_name, name=cv_name)
+            self.assertIsNotNone(
+                self.content_views.search(cv_name)
+            )
+            # Add repository to selected CV
+            self.content_views.add_remove_repos(cv_name, [repo_name])
+            self.assertIsNotNone(
+                self.content_views.wait_until_element(
+                    common_locators["alert.success"])
```
Comment (Ichimonji10) [inline tests/foreman/ui/test_contentviews.py]: Nitpick: `s/"/'/g`

File: robottelo/ui/contentviews.py [hunk 14]
```diff
@@ -722,3 +722,41 @@ def fetch_puppet_module(self, cv_name, module_name):
...
+        else:
+            raise UIError('Could not copy the Content View %s .', name)
+
+    def fetch_yum_content_repo_name(self, cv_name):
+        """Fetch associated content host from selected activation key."""
+        # find content_view
+        cv = self.search(cv_name)
+        if cv is None:
+            raise UINoSuchElementError(
+                'Could not find CV %s', cv_name)
```
Comment (Ichimonji10) [inline robottelo/ui/contentviews.py]: Will this fit on one line?

File: tests/foreman/ui/test_contentviews.py [hunk 15]
```diff
@@ -977,6 +976,37 @@ def test_cv_clone_within_same_env(self):
...
+        repo_name = gen_string('alpha', 8)
+        cv_name = gen_string('alpha', 8)
+        copy_cv_name = gen_string('alpha', 8)
+        with Session(self.browser) as session:
+            self.setup_to_create_cv(repo_name=repo_name)
+            # Create content-view
+            make_contentview(session, org=self.org_name, name=cv_name)
+            self.assertIsNotNone(
+                self.content_views.search(cv_name)
+            )
```
Comment (Ichimonji10) [inline tests/foreman/ui/test_contentviews.py]: Fits on one line?

File: tests/foreman/ui/test_contentviews.py [hunk 16]
```diff
@@ -977,6 +976,37 @@ def test_cv_clone_within_same_env(self):
...
+            self.content_views.publish(cv_name)
+            self.assertIsNotNone(
+                self.content_views.wait_until_element(
+                    common_locators["alert.success"])
+            )
+            # Copy the CV
+            self.content_views.copy_view(cv_name, copy_cv_name)
+            self.assertIsNotNone(
+                self.content_views.search(copy_cv_name)
+            )
```
Comment (Ichimonji10) [inline tests/foreman/ui/test_contentviews.py]: Fits on one line?

**Comment by Ichimonji10** [comment]
The code has been significantly improved. Thanks for the work, @jyejare. ACK pending nitpicks.

File: robottelo/ui/contentviews.py (hunk 14 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Put on same line.

File: tests/foreman/ui/test_contentviews.py (hunk 15 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Put on same line.

File: tests/foreman/ui/test_contentviews.py (hunk 16 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Put on one line.

File: robottelo/ui/contentviews.py (hunk 7 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Done

File: robottelo/ui/contentviews.py (hunk 8 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Done

File: robottelo/ui/contentviews.py (hunk 10 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Done

File: robottelo/ui/contentviews.py (hunk 9 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Done

File: robottelo/ui/contentviews.py (hunk 11 above)
Comment (jyejare) [inline robottelo/ui/contentviews.py]: Done.

File: tests/foreman/ui/test_contentviews.py (hunk 13 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Done.

File: tests/foreman/ui/test_contentviews.py (hunk 12 above)
Comment (jyejare) [inline tests/foreman/ui/test_contentviews.py]: Done.

**Comment by jyejare** [comment]
@Ichimonji10 and @elyezer  , again the comments has been taken care. Please review again to merge. Thanks again for this learning curve.

**Comment by elyezer** [comment]
APT

**Comment by Ichimonji10** [comment]
ACK

**Pull request merged**